ztp/source-crs/optional-extra-manifest/ipsec/build.sh

```diff
@@ -0,0 +1,75 @@
+#!/bin/bash
```

Comment:
ztp won't execute this sh asaik. @lack @sabbir-47 could you confirm? 

So the MC yaml should be added to source-cr directly instead.

Reply:
Maybe the build.sh can go here if we really want to create MC files using script. https://github.com/openshift-kni/cnf-features-deploy/tree/master/ztp/extra-manifests-builder

Reply:
The script generates a MC with the certificates to import in base64, since the certificates are changing between our partners and our setups it can't be a static MC.
Please follow the README and the official documentation https://docs.openshift.com/container-platform/4.15/networking/ovn_kubernetes_network_provider/configuring-ipsec-ovn.html#nw-ovn-ipsec-external_configuring-ipsec-ovn 
You should run it as a pre step to generate the MC that will be applied with ZTP. You can run that in your automation as a pre step to ZTP.

Reply:
@sabinaaledort The ipsec folder should move to extra-manifests-builder: https://github.com/openshift-kni/cnf-features-deploy/tree/master/ztp/extra-manifests-builder. The scripts should be as much automatic as possible and the static part should be highlighted.

Reply:
The script generates a MC with the certificates to import in base64, since the certificates are changing between our partners and our setups it can't be a static MC.
The parts to change are described in the README and the official documentation https://docs.openshift.com/container-platform/4.15/networking/ovn_kubernetes_network_provider/configuring-ipsec-ovn.html#nw-ovn-ipsec-external_configuring-ipsec-ovn
This is a pre step to do once or as an automation in the pipeline before ZTP.
This is the path we agreed to store the files since this is an optional feature. If that's the day 0 path we can move it to the path you suggested.

Reply:
Something to keep in mind here:
In customer/partner deployments, there will be a unique MC generated per target far-edge SNO deployment. This is because a unique IPSec certificate is needed per SNO cluster and the ipsec-endpoint-config.yml file will have a unique SNO IP per SNO cluster.
Also, the IPSec config and cert should be applied as a Day2 operation. This is because usually these SNO clusters will first be deployed with OCP, etc (many times in a factory environment) without IPSec, then post-deployment the IPSec tunnel will be configured and started.

Reply:
@sabinaaledort everything in source-crs folder is meant to be consumable as part of the ZTP process directly regardless of core features or optional features.

The "extra-manifests" subfolders are for day0 config, and the rest are for day2 configs. User should be able to use/reference them from user git repo in siteconfig or in PGT.
 
For example BMER is an optional feature, but all the BMER related source-crs stay in this folder because its user's choice on whether to use/reference them in day2 PGT at the end.

If the solution requires prereq steps prior to ZTP, it needs to be documented separately. So far, I've seen most of the readmes go into this folder: https://github.com/openshift-kni/cnf-features-deploy/tree/master/ztp/gitops-subscriptions/argocd. Maybe @sabbir-47 can suggest something better if any. 

@sabbir-47 , it seems the sh requires 2 external cert files for each site as @bradyallenjohnson mentioned, I guess in this case, there is no easy way to integrate this script to extra-manifests-builder in this case?

Reply:
I think, the CRs which are directly consumable should go under source-crs directory, in this case it is not.

> If the solution requires prereq steps prior to ZTP, it needs to be documented separately. So far, I've seen most of the readmes go into this folder: https://github.com/openshift-kni/cnf-features-deploy/tree/master/ztp/gitops-subscriptions/argocd. Maybe @sabbir-47 can suggest something better if any.

Readme must go there.

> @sabbir-47 , it seems the sh requires 2 external cert files for each site as @bradyallenjohnson mentioned, I guess there is no easy way to integrate this script to extra-manifests-builder in this case?

We don't provide a way to integrate scripts directly into ztp, apart from the ones that can be build using exta-manifests-builder which doesn't depend on the user input.

Reply:
> I think, the CRs which are directly consumable should go under source-crs directory, in this case it is not.

Keep in mind also that many of the source-crs already present are NOT directly applicable to user systems; many of them must have at least some parts overridden in PGTs.

> > If the solution requires prereq steps prior to ZTP, it needs to be documented separately. So far, I've seen most of the readmes go into this folder: https://github.com/openshift-kni/cnf-features-deploy/tree/master/ztp/gitops-subscriptions/argocd. Maybe @sabbir-47 can suggest something better if any.
> 
> Readme must go there.

I disagree about the readme.  This readme is specific to this set of yamls and the related build.sh - I think it makes more sense for them all to be together in their own subdirectory.

> > @sabbir-47 , it seems the sh requires 2 external cert files for each site as @bradyallenjohnson mentioned, I guess there is no easy way to integrate this script to extra-manifests-builder in this case?
> 
> We don't provide a way to integrate scripts directly into ztp, apart from the ones that can be build using exta-manifests-builder which doesn't depend on the user input.

I agree that this cannot and should not go in extra-manafifests-builder.  That is a source-control of mechanism to build static MachineConfigs that end up in source-crs/extra-manifests.  That doesn't fit this use-case, as these are not static yaml files.

Reply:
> The build.sh script should not be included in ztp. It should only be used as a pre-step by the partner/customer to generate the MC.

The build.sh script is a critical part of the customer-facing solution.  It simplifies creation of yaml files that are essential to customers implementing this feature.  I think that is exactly why it belongs in the ZTP container.

Reply:
Ok, perhaps this is a question of nomenclature. The customers will not be executing this script on the SNO cluster, but from a staging env where they launch the SNO deployments. When I say it should not be in ZTP, I mean ZTP should not install the script on the SNO cluster.

Reply:
> ZTP should not install the script on the SNO cluster.

Agreed!

Reply:
So it seems to me like we have 2 viable options here, armed with the knowledge that these files are not going to be installed directly on an SNO by default:
- Keep these files where they are under source-crs/optional-extra-manifests/ipsec
- Move the files into the "examples" subdirectory alongside the example siteconfig and example PGTs we already ship in the ZTP container

Reply:
@lack 
> Keep in mind also that many of the source-crs already present are NOT directly applicable to user systems; many of them must have at least some parts overridden in PGTs.

Yes, we can override in some of the provided source-crs, e.g. performanceProfile, tuned, sriov etc. But in this case, can a user override some parameter and consume the yaml as it is? I don't think so.

I would be up for moving the ipsec under the examples directory documenting how a user is supposed to use it and the steps.